Add support for bandlimited angular spectrum.

ebezzam opened the pull request:
@GJBoth rough example and implementation to add support for BLAS (bandlimited angular spectrum)

References:
- original paper: https://opg.optica.org/oe/fulltext.cfm?uri=oe-17-22-19662&id=186848#g002
- off-axis: https://opg.optica.org/oe/fulltext.cfm?uri=oe-18-17-18453&id=205150#articleTables

I've tried to create an example that replicates Fig. 9a of the [original paper](https://opg.optica.org/oe/fulltext.cfm?uri=oe-17-22-19662&id=186848#g002). It's not exactly the same (but close). We do see the high-frequency noise found in the original AS being suppressed.

![propagation_comparison](https://github.com/chromatix-team/chromatix/assets/4757445/372362df-e6a8-4879-8fb7-997db7390c0e)

TODO:
- [ ] implement direct/numerical integration as a ground-truth for comparison, as they do in Fig 4 and 6 of [original paper](https://opg.optica.org/oe/fulltext.cfm?uri=oe-17-22-19662&id=186848#g002). Example implementation [here](https://github.com/ebezzam/waveprop/blob/a2d65116336bfb6e95732fd982e5c3ec2109cff3/waveprop/rs.py#L33)
- [ ] verify off-axis

ebezzam commented on src/chromatix/functional/propagation.py:
```diff
@@ -266,7 +271,62 @@ def compute_asm_propagator(
+        # bandlimit
+        H_filter = (np.abs(fX - u0) <= fx_max) * (np.abs(fY - v0) < fy_max)
+
+        # to jax
```
probably a better way to do above with Jax?

diptodip replied:
This is making a rectangle mask in the frequency grid? If so we shouldn't need `jnp.where` and this should just work because it's a binary mask and not variably sized (assuming the lines above can have their conditionals removed).

ebezzam replied:
Yes it's making a rectangular mask as shown in [Fig 5](https://opg.optica.org/getimagev2.cfm?img=7dKLEDezXUrkoUkzVOSYXrMOHiBaTdFKRs1SIJ54Pfo%3D&size=full&uri=oe-18-17-18453-g005) of the off-axis paper. I'll see if there's a way to remove the conditions before.

ebezzam replied:
done!

GJBoth commented on src/chromatix/functional/propagation.py:
```diff
@@ -266,7 +271,62 @@ def compute_asm_propagator(
+        du = 1 / (2 * Sx)
+        u_limit_p = ((x0 + 1 / (2 * du)) ** (-2) * z0**2 + 1) ** (-1 / 2) / wv
+        u_limit_n = ((x0 - 1 / (2 * du)) ** (-2) * z0**2 + 1) ** (-1 / 2) / wv
+        if Sx < x0:
+            u0 = (u_limit_p + u_limit_n) / 2
+            u_width = u_limit_p - u_limit_n
+        elif x0 <= -Sx:
+            u0 = -(u_limit_p + u_limit_n) / 2
+            u_width = u_limit_n - u_limit_p
+        else:
+            u0 = (u_limit_p - u_limit_n) / 2
+            u_width = u_limit_p + u_limit_n
```
We have to be careful here - jax _really_ doesn't like conditionals based on values, so we shouldn't do this. Is there some way to write this without if/else?

diptodip replied:
Maybe using `jnp.sign(x0 - Sx)` or something like that?

ebezzam replied:
yep I think applying `jnp.sign` correctly does the trick

GJBoth commented on src/chromatix/functional/propagation.py:
```diff
@@ -266,7 +271,62 @@ def compute_asm_propagator(
+        dv = 1 / (2 * Sy)
+        v_limit_p = ((y0 + 1 / (2 * dv)) ** (-2) * z0**2 + 1) ** (-1 / 2) / wv
+        v_limit_n = ((y0 - 1 / (2 * dv)) ** (-2) * z0**2 + 1) ** (-1 / 2) / wv
+        if Sy < y0:
+            v0 = (v_limit_p + v_limit_n) / 2
+            v_width = v_limit_p - v_limit_n
+        elif y0 <= -Sy:
+            v0 = -(v_limit_p + v_limit_n) / 2
+            v_width = v_limit_n - v_limit_p
+        else:
+            v0 = (v_limit_p - v_limit_n) / 2
+            v_width = v_limit_p + v_limit_n
```
Would there be some way to do for this x and y at the same time by vectorizing it?

ebezzam replied:
done!

ebezzam commented:
@GJBoth and @diptodip thanks for your comments! I'll take a look.

One dummy question: is [off-axis](https://opg.optica.org/oe/fulltext.cfm?uri=oe-18-17-18453&id=205150) really the same as your [kykx](https://github.com/chromatix-team/chromatix/blob/7304cd312b28eebc2f15c3c466e53074141d553b/src/chromatix/functional/propagation.py#L162)?

From what I understand `kykx` is tilting the plane wave illumination (which can shift the diffraction pattern like in the top row of Fig 7 of the [paper](https://opg.optica.org/oe/fulltext.cfm?uri=oe-18-17-18453&id=205150)), while off-axis is interested in sampling the destination field *off-axis* (Fig 1 and bottom row of Fig 7 in the [paper](https://opg.optica.org/oe/fulltext.cfm?uri=oe-18-17-18453&id=205150)).

ebezzam commented on src/chromatix/functional/propagation.py:
```diff
@@ -266,7 +271,62 @@ def compute_asm_propagator(
+        dfY = 1.0 / Sy
+        N_y, N_x = field.spatial_shape
+        fX = np.linspace(-N_x // 2, N_x // 2 - 1, num=N_x)[np.newaxis, :] * dfX
+        fY = np.linspace(-N_y // 2, N_y // 2 - 1, num=N_y)[:, np.newaxis] * dfY
```
done!

ebezzam commented:
@GJBoth and @diptodip I think I've addressed your comments! It's now vectorized for y and x and without if-else conditions. I've left the original code commented because I wonder if it's easier for someone to read that than the current, very dense version?

Also I found a small bug in the previous version (I had an extra factor 2 because in the paper they talked about the field size being increased by 2x to linearize the convolution, but within your propagator function the field is already padded). So now the result looks more like the original paper.

![propagation_comparison](https://github.com/chromatix-team/chromatix/assets/4757445/c080e6d0-ff49-4455-9b75-ae5bb47d0253)

Below is from original paper (maybe they apply a gamma correction?)

<img width="326" alt="Screenshot 2024-05-25 at 6 02 01 PM" src="https://github.com/chromatix-team/chromatix/assets/4757445/cb792545-c87f-45ad-a54c-7b55a15a28d6">

As mentioned above, I think off-axis may be different from your kykx for tilting the illumination? Can you confirm? If so, I think it could be interesting to add a `shift_yx` optional argument for shifting the destination sampling window as done in the off-axis paper.

diptodip commented on examples/bandlimited_angular_spectrum.py:
This is a great example but would you mind to move it to a notebook with the plots included under `docs/examples/bandlimited_angular_spectrum.ipynb`? I can do it myself so it's no big deal but that way people can see the plots and we can generate a page on the docs website for it.

diptodip commented on src/chromatix/functional/propagation.py:
So like you said, shifting is different from just the `kykx` and like Rainer pointed out I think the best thing to do is to add a property to the `Field` to keep track of this. But we can do that later.

And I think it's better to get rid of the comments :)

I don't think the code is that dense and we can of course add some explanation comments of how the coordinates are being chosen.

And finally (not a big deal but just a note for myself later) we have `l1_norm` in the utilities that we use in `square_pupil` to make rectangular masks.

diptodip commented on src/chromatix/field.py:
Great add!

diptodip commented:
@ebezzam This all looks awesome! Sorry I'm just now getting to see this. By the way, I'm going to update this PR to merge into `hackathon_2024` instead of `main`. I'll leave some more specific comments but in general:

> @GJBoth and @diptodip thanks for your comments! I'll take a look.
> 
> One dummy question: is [off-axis](https://opg.optica.org/oe/fulltext.cfm?uri=oe-18-17-18453&id=205150) really the same as your [kykx](https://github.com/chromatix-team/chromatix/blob/7304cd312b28eebc2f15c3c466e53074141d553b/src/chromatix/functional/propagation.py#L162)?
> 
> From what I understand `kykx` is tilting the plane wave illumination (which can shift the diffraction pattern like in the top row of Fig 7 of the [paper](https://opg.optica.org/oe/fulltext.cfm?uri=oe-18-17-18453&id=205150)), while off-axis is interested in sampling the destination field _off-axis_ (Fig 1 and bottom row of Fig 7 in the [paper](https://opg.optica.org/oe/fulltext.cfm?uri=oe-18-17-18453&id=205150)).

Yes I think you're right about that. And Rainer also made the great point today that we could simply add a new property to our `Field` objects to keep track of a spatial location (destination sampling window  center as in the paper you shared). This way when we propagate off axis we can just update this property instead of requiring propagation of a large field of view (of course both could still be allowed).

So this I think is definitely a nice thing that we should add, but perhaps we work on that update in another PR?

> Below is from original paper (maybe they apply a gamma correction?)
> 
> <img alt="Screenshot 2024-05-25 at 6 02 01 PM" width="326" src="https://private-user-images.githubusercontent.com/4757445/333816314-cb792545-c87f-45ad-a54c-7b55a15a28d6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY2OTQ3NzAsIm5iZiI6MTcxNjY5NDQ3MCwicGF0aCI6Ii80NzU3NDQ1LzMzMzgxNjMxNC1jYjc5MjU0NS1jODdmLTQ1YWQtYTU0Yy03YjU1YTE1YTI4ZDYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUyNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MjZUMDMzNDMwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ODdlODBlMmZjZjRiM2IzZWM2YmM2YWVkNzg1YmEzMDgzZDljOTJmNGJmZGNiZTRlNjg5MDljMTM1ZDQwOTk3MSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.SZWQ_3duu3yp7kR7hm9nnpIrrRYGcgNeOpQr1F8TZM4">

It does look like a gamma correction. What does it look like if you plot your results with a gamma correction (e.g. `intensity**0.7` or so)? Or they plot amplitude and not intensity?

ebezzam commented:
@diptodip thanks for your comments! I'll address them on Monday

ebezzam commented:
@diptodip, @GJBoth I've addressed your comments, thanks for your support!

I've created a notebook which shows BLAS and off-axis propagation:

<img width="585" alt="Screenshot 2024-05-28 at 3 02 41 PM" src="https://github.com/chromatix-team/chromatix/assets/4757445/9c5e2d22-ab06-4415-9770-f0dd9432eada">


<img width="581" alt="Screenshot 2024-05-28 at 3 03 02 PM" src="https://github.com/chromatix-team/chromatix/assets/4757445/839d6c31-68f1-4e40-8a37-5074ca1d8165">

I've added a `shift_yx` optional parameter to `compute_asm_propagator` so that there is a working implementation of off-axis, but if you think it's better to move that to the `Field` class, it should be straightforward.

Let me know if you need anything else!

diptodip commented:
Thanks Eric! We'll merge things into the hackathon_2024 branch and go from there. Really appreciate all your work implementing this!

RainerHeintzmann commented:
Cool stuff you added! Looks great. Also the off-axis. Nice.
